File: python/1448.py
from data_strcutres import TreeNode

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def dfs(node, vmax):
            count = 0
            if node.val >= vmax:
                count += 1
            vmax = max(vmax, node.val)
            if node.left:
                count += dfs(node.left, vmax)
            if node.right:
                count += dfs(node.right, vmax)
            return count
        return dfs(root, root.val)

File: cpp/1448.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int goodNodes(TreeNode* root) {
        return dfs(root, root->val);
    }
private:
    int dfs(TreeNode* node, int vmax) {
        int count = 0;
        if (node->val >= vmax) {
            count++;
            vmax = node->val;
        }
        if (node->left != nullptr) {
            count += dfs(node->left, vmax);
        }
        if (node->right != nullptr) {
            count += dfs(node->right, vmax);
        }
        return count;
    }
};
